### PR TITLE
[LIVE-2006][LIVE-2007] Feature: EIP-1559

### PR DESCRIPTION
**Pull request by lambertkevin**
### 📝 Description

Implementing EIP-1559 for the Ethereum family transactions.
To the extent of this scope, this PR is cleaning a little bit the Ethereum coin integration and splitting its bridge into multiple files.

### ❓ Context

- **Impacted projects**: `live-common` `ledger-live-desktop` `ledger-live-mobile` `ledgerjs` 
- **Linked resource(s)**: `https://ledgerhq.atlassian.net/browse/LIVE-2006` `https://ledgerhq.atlassian.net/browse/LIVE-2007` `` 

### ✅ Checklist

- [x] **Test coverage**  // Test coverage is only done with the bot. 😞 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

**Comment by gre** [comment]
you may want to rebase one more time :) @lambertkevin

**Comment by lambertkevin** [comment]
P

> ## ⏲ 16min 34s ✅ 4 txs ❌ 2 txs ($11.44) for [**[Bot] Testing with 'Mooncake'**](https://github.com/LedgerHQ/ledger-live/actions/runs/3389414131)
> > ✅ 2 specs are successful: _Polygon, Ethereum Goerli_
> > ❌ 1 specs have problems: _BSC_
> > 💰 1 specs may miss funds: _Ethereum Classic_
> 
> > What is the bot and how does it work? [Everything is documented here!](https://github.com/LedgerHQ/ledger-live/wiki/LLC:bot)
> 
> ❌ 2 mutation errors
> ⚠️ 5 spec hints
> Details of the 6 mutations
> Details of the 9 uncovered mutations
> Portfolio ($11.44) – Details of the 5 currencies
> Performance ⏲ 16min 34s
> > What is the bot and how does it work? [Everything is documented here!](https://github.com/LedgerHQ/ledger-live/wiki/LLC:bot)

Probably related to the BNB explorer incident ? 🤔

File: apps/ledger-live-mobile/src/icons/TachometerFast.tsx [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+import Svg, { Path } from "react-native-svg";
+
+type Props = {
```
Comment (lvndry) [inline apps/ledger-live-mobile/src/icons/TachometerFast.tsx:4]: 2 questions:

- Why making these props mandatory as they have default values in the component
- Couldn't we make this a shared `IconProps` used by all icons ?

File: libs/ledger-live-common/src/env.ts [hunk 2]
```diff
@@ -46,6 +46,11 @@ const jsonParser = (v: unknown): JSONValue | undefined => {
   }
 };
 
+const stringArrayParser = (v: any): string[] | null | undefined => {
```
Comment (lvndry) [inline libs/ledger-live-common/src/env.ts:49]: Function can't return undefined

Comment (lvndry) [reply libs/ledger-live-common/src/env.ts:49]: Why not forcing the type of `v` to be `string` here ?

File: libs/ledger-live-common/src/families/ethereum/logic.ts [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export const isEthereumAddress = (address: string): boolean =>
+  Boolean(address.match(/^0x[0-9a-fA-F]{40}$/));
```
Comment (lvndry) [inline libs/ledger-live-common/src/families/ethereum/logic.ts:2]:
```suggestion
  /^0x[0-9a-fA-F]{40}$/.test(address);
```

File: libs/ledgerjs/packages/cryptoassets/src/currencies.ts [hunk 4]
```diff
@@ -3093,7 +3093,7 @@ export const cryptocurrenciesById: Record<string, CryptoCurrency> = {
       baseChain: "goerli",
       chainId: 5, // goerli
       networkId: 5,
-      hardfork: "petersburg",
+      hardfork: "london",
```
Comment (lvndry) [inline libs/ledgerjs/packages/cryptoassets/src/currencies.ts:3096]: - Why `london` specifically and not a more recent fork ?
- Why `ethereum_ropsten` is still using `petersburg` ?

File: apps/ledger-live-desktop/src/renderer/families/ethereum/GasLimitField.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+type Props = {
+  transaction: Transaction;
+  account: Account;
+  status: TransactionStatus;
+  updateTransaction: Result["updateTransaction"];
+};
+
+const AdvancedOptions = ({ account, transaction, status, updateTransaction }: Props) => {
+  invariant(transaction.family === "ethereum", "AdvancedOptions: ethereum family expected");
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/families/ethereum/GasLimitField.tsx:23]: Invariant will display an error

File: libs/ledgerjs/packages/cryptoassets/src/currencies.ts (hunk 4 above)
Comment (lambertkevin) [reply libs/ledgerjs/packages/cryptoassets/src/currencies.ts:3096]: Because EIP-1559 has been introduced with the London fork, so we're supporting up until this point.
Ropsten isn't supported anymore in the Live.

File: apps/ledger-live-desktop/src/renderer/families/ethereum/SendAmountFields.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+type Props = {
+  transaction: EthereumTransaction;
+  account: AccountLike;
+  parentAccount: Account | undefined;
+  updateTransaction: Result<EthereumTransaction>["updateTransaction"];
+  status: TransactionStatus;
+};
+
+const Root = (props: Props) => {
+  const { account, parentAccount, updateTransaction, transaction } = props;
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/families/ethereum/SendAmountFields.tsx:30]: Nope, `props` is spread to all the fees related components right after.

File: apps/ledger-live-desktop/src/renderer/components/Box/Box.js [hunk 7]
```diff
@@ -35,11 +35,9 @@ const Box: ThemedComponent<{
   relative?: boolean,
   sticky?: boolean,
   selectable?: boolean,
-  // DROP this when we fully migrated from V1
-  align?: "THIS PROPERTY IS NOT VALID – SWITCH TO alignItems",
-  justify?: "THIS PROPERTY IS NOT VALID – SWITCH TO justifyContent",
```
Comment (gre) [inline apps/ledger-live-desktop/src/renderer/components/Box/Box.js:40]: unclear why this is related to this PR. maybe send a separate PR?

File: .github/workflows/bot-ethereum-phosphore.yml [hunk 8]
```diff
@@ -0,0 +1,54 @@
+name: "[Bot] 'Ethereum on Phosphore'"
```
Comment (gre) [inline .github/workflows/bot-ethereum-phosphore.yml:1]: please remove this file, we stopped doing these kind of config files. you can directly use bot-phosphore.yml to trigger flows.

File: apps/ledger-live-mobile/src/families/ethereum/EthereumCustomFees/Ethereum1559CustomFees.tsx [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+  transaction: originalTransaction,
+}: Props) => {
+  const { colors } = useTheme();
+
+  invariant(originalTransaction.family === "ethereum", "not ethereum family");
+  invariant(account, "no account found");
+
+  const { currency } = getMainAccount(account, parentAccount);
+  const feeUnit =
+    currency.units.length > 1 ? currency.units[1] : currency.units[0];
```
Comment (gre) [inline apps/ledger-live-mobile/src/families/ethereum/EthereumCustomFees/Ethereum1559CustomFees.tsx:41]: it could be wise to externalize this logic somewhere else, like "getDefaultFeeUnit" on ethereum helper. because it's weird why [1] and [0] (and i think we may have this in multiple places)

File: apps/ledger-live-mobile/src/icons/TachometerFast.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  size: number;
+  color: string;
+};
+
+export default function TachometerFast({
```
Comment (gre) [inline apps/ledger-live-mobile/src/icons/TachometerFast.tsx:9]: React.memo could be good. also i wonder if it should be in the "ui" project

**Review by gre — APPROVED** [review]
some feedbacks. overall is good. general feedback is it would be valuable to extract out some of the non related work that is outside of families/* folder into a dedicated PR on which you depend on.

File: libs/ledger-live-common/src/env.ts (hunk 2 above)
Comment (lambertkevin) [reply libs/ledger-live-common/src/env.ts:49]: Part of the code that I got from the previous PR, but with the way it's coded right now, yes it can return `undefined`.
```typescript
if (Array.isArray(v_array) && v_array.length > 0) return v_array;
// else return undefined
```

for the `v` type, `envDefinitions` is typed as `Record<EnvName, EnvDef<any>>` where `EnvDef` is:
```
type EnvDef<V> = {
  desc: string;
  def: V;
  parser: (arg0: unknown) => V | null | undefined;
};
``` 
Meaning that at this stage, the typing is poor, but without changing everything (since it's not related to this PR) `v` can be either `unknown` or `any`.

File: apps/ledger-live-desktop/src/renderer/components/Box/Box.js (hunk 7 above)
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/components/Box/Box.js:40]: Got this from the previous PR, will remove yes 👌

File: apps/ledger-live-mobile/src/icons/TachometerFast.tsx (hunk 1 above)
Comment (lambertkevin) [reply apps/ledger-live-mobile/src/icons/TachometerFast.tsx:4]: - Added with the memoization 👌 
- That would be a live-hub topic, I don't know how feasible or not this is, especially now that we're building new icons library I think.
